### PR TITLE
[WIP] #40: Documentation Standards

### PR DESCRIPTION
**Pull request by mvriel**
I have created a baseline of handling Documentation Standards. phpDocumentor now supports rulesets in the same way as PHP_CodeSniffer and as such allows the custom creation of standards.

With this feature it will be possible to finetune and tweak how your documentation is validated.

This PR is still a work in progress as several features have not been added yet:

- [ ] Setting a Standard from the command line by name
- [ ] Setting a Standard from the command line by file name (separate ruleset.xml)
- [ ] Read a Standard from the phpdoc.xml
- [ ] User documentation
- [ ] Fully covered by unit tests
- [ ] Move basic standards package to a Components folder instead of Plugins

### CODE REVIEW

File: src/phpDocumentor/Plugin/Standards/RulesetLoader.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        if (!isset($this->rulesets[$rulesetName])) {
+            throw new \InvalidArgumentException('Ruleset with the name "' . $rulesetName . '" was not found');
+        }
+
+        $ruleset = $this->rulesets[$rulesetName];
+        $ruleset->enableSniffs($this->sniffs);
+
+        return $ruleset;
+    }
+}
```
Comment (jaapio) [inline src/phpDocumentor/Plugin/Standards/RulesetLoader.php]: missing a new line here

**Comment by mvriel** [comment]
*Important*: due to the changes in phpDocumentor 3 it will be rather difficult to finish this specific PR; once the new validation component for phpDocumentor 3 is designed we can take this effort into account and possibly re-use parts of this.

Due to the above I will keep this PR open until Validation is built but after that close it

File: composer.json [hunk 2]
```diff
@@ -7,7 +7,7 @@
     "autoload": {
         "psr-0":{
```
Comment (rgeraads) [inline composer.json]: would it be an idea to switch to psr-4 autoloading now that changes are made here, to make it future proof?

File: src/phpDocumentor/Plugin/Standards/Ruleset.php [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+ * A collection of Rules that together form a Documentation Standard.
+ *
+ * A Ruleset is a series of Rules that determine what properties of a Project are 'sniffed', or checked. A Ruleset may
+ * also exclude a series of paths from the checking by child Rules (and subsequent Child Rulesets references by those
+ * Rules).
+ *
+ * @todo do not enable Sniffs that are excluded in a Rule
+ * @todo apply exclusion rules to child Rules.
+ */
+class Ruleset
```
Comment (rgeraads) [inline src/phpDocumentor/Plugin/Standards/Ruleset.php]: doesn't this need to be a final class?

File: src/phpDocumentor/Plugin/Standards/ServiceProvider.php [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * This service provider depends on the Symfony Validator (included in the Symfony Validation Component) because a
+ * Sniff registers a Validator on a specific class (for phpDocumentor this is always a Descriptor). By invoking the
+ * Validator after a Ruleset is loaded we can verify any object.
+ *
+ * The Violations that result from Validators created by Sniffs will always contain the name of the Rule that enabled
+ * the Sniff. The calling application can then search for the associated Rule to determine its Message and Severity
+ * using the {@see Ruleset::getRule()) method.
+ */
+class ServiceProvider implements ServiceProviderInterface
```
Comment (rgeraads) [inline src/phpDocumentor/Plugin/Standards/ServiceProvider.php]: final class?

File: tests/unit/phpDocumentor/Plugin/Standards/phpDocumentor/Sniff/ArgumentHasParamTagTest.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    /** @var ArgumentHasParamTag */
+    private $fixture;
+
+    /**
+     * Initializes the fixture and its dependencies with mocks.
+     */
+    protected function setUp()
+    {
+        $this->validatorMock = m::mock('Symfony\Component\Validator\Validator');
```
Comment (rgeraads) [inline tests/unit/phpDocumentor/Plugin/Standards/phpDocumentor/Sniff/ArgumentHasParamTagTest.php]: you might want to change all the hardcoded class strings to e.g. ```validator::class``` to allow code navigation and see usage statistics

**Pull request closed without merging**